CMake/FBThriftCppLibrary.cmake

```diff
@@ -0,0 +1,161 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+
+include(FBCMakeParseArgs)
+
+# Generate a C++ library from a thrift file
+#
+# Parameters: - SERVICES <svc1> [<svc2> ...] The names of the services defined
+# in the thrift file. - DEPENDS <dep1> [<dep2> ...] A list of other thrift C++
+# libraries that this library depends on. - OPTIONS <opt1> [<opt2> ...] A list
+# of options to pass to the thrift compiler. - INCLUDE_DIR <path> The
+# sub-directory where generated headers will be installed. Defaults to "include"
+# if not specified.  The caller must still call install() to install the thrift
+# library if desired. - THRIFT_INCLUDE_DIR <path> The sub-directory where
+# generated headers will be installed. Defaults to "${INCLUDE_DIR}/thrift-files"
+# if not specified. The caller must still call install() to install the thrift
+# library if desired.
+function(add_fbthrift_cpp_library LIB_NAME THRIFT_FILE)
+  # Parse the arguments
+  set(one_value_args INCLUDE_DIR THRIFT_INCLUDE_DIR)
+  set(multi_value_args SERVICES DEPENDS OPTIONS)
+  fb_cmake_parse_args(ARG "" "${one_value_args}" "${multi_value_args}"
+                      "${ARGN}")
+  if(NOT DEFINED ARG_INCLUDE_DIR)
+    set(ARG_INCLUDE_DIR "include")
+  endif()
+  if(NOT DEFINED ARG_THRIFT_INCLUDE_DIR)
+    set(ARG_THRIFT_INCLUDE_DIR "${ARG_INCLUDE_DIR}/thrift-files")
+  endif()
+
+  get_filename_component(base ${THRIFT_FILE} NAME_WE)
+  get_filename_component(output_dir ${CMAKE_CURRENT_BINARY_DIR}/${THRIFT_FILE}
+                         DIRECTORY)
+
+  # Generate relative paths in #includes
+  file(RELATIVE_PATH include_prefix "${CMAKE_SOURCE_DIR}"
+       "${CMAKE_CURRENT_SOURCE_DIR}/${THRIFT_FILE}")
+  get_filename_component(include_prefix ${include_prefix} DIRECTORY)
+
+  if(NOT "${include_prefix}" STREQUAL "")
+    list(APPEND ARG_OPTIONS "include_prefix=${include_prefix}")
+  endif()
+  # CMake 3.12 is finally getting a list(JOIN) function, but until then treating
+  # the list as a string and replacing the semicolons is good enough.
+  string(REPLACE ";" "," GEN_ARG_STR "${ARG_OPTIONS}")
+
+  # Compute the list of generated files
+  list(
+    APPEND
+    generated_headers
+    "${output_dir}/gen-cpp2/${base}_constants.h"
+    "${output_dir}/gen-cpp2/${base}_types.h"
+    "${output_dir}/gen-cpp2/${base}_types.tcc"
+    "${output_dir}/gen-cpp2/${base}_types_custom_protocol.h"
+    "${output_dir}/gen-cpp2/${base}_metadata.h")
+  list(
+    APPEND
+    generated_sources
+    "${output_dir}/gen-cpp2/${base}_constants.cpp"
```

Comment:
nit: curious - why not merge these 2 calls into one call to list APPEND ?

Reply:
they create two different lists? Though this is the code I vendored from fbthrift (I haven't looked at the logic very closely)

Reply:
Ah Its generated sources/headers - missed that..